### PR TITLE
fix: add a ClusterName property and use it for resources names

### LINKED ISSUE #56: Enable user-configurable vm names
Details // TODO

Some customers wish to be able to influence the names of their node vms. Scale/upgrade would have to be taken into account in order to deliver this.

### PR DESCRIPTION
**Pull request by sylr**
**What this PR does / why we need it**:

Rebase a branch we used at my company to fix the name of the resources.

It adds a property `clusterName` which is used to prefix the name of all Azure resources, e.g. : `clusterName = k8s-prod` -> `k8s-prod-master-0, k8s-prod-master-1 ... etc`

If no  `clusterName` is provided the prefix of the resources falls back to `(k8s|aks)-<clusterId>`.

**Which issue this PR fixes** : fixes #56

**Special notes for your reviewer**:

**Release note**:

```release-note
Add a clusterName property which is used to prefix all Azure resources. 
```

### CODE REVIEW

**Comment by sylr** [comment]
Yes. I rebased this branch onto #202 to be able to run tests.

**Comment by sylr** [comment]
@tariq1890 can you approve the e2e checks on circleci, I'll remove the goimports patch tomorrow.

File: pkg/helpers/helpers.go [hunk 1]
```diff
@@ -140,7 +140,7 @@ func AcceleratedNetworkingSupported(sku string) bool {
 		"Standard_D64_v3", "Standard_D64s_v3", "Standard_E64_v3", "Standard_E64s_v3", "Standard_E64-16s_v3",
 		"Standard_E64-32s_v3", "Standard_F64s_v2", "Standard_F72s_v2", "Standard_M128s", "Standard_M128ms",
 		"Standard_L8s_v2", "Standard_L16s_v2", "Standard_L32s_v2", "Standard_L64s_v2", "Standard_L96s_v2",
-		"SQLGL", "SQLGLCore", "Standard_D4_v3", "Standard_D4s_v3", "Standard_D2_v2", "Standard_DS2_v2",
+		"SQLGL", "SQLGLCore", "Standard_D4_v3", "Standard_D4s_v3", "Standard_D2s", "Standard_D2_v2", "Standard_DS2_v2", "Standard_D2s_v3",
```
Comment (tariq1890) [inline pkg/helpers/helpers.go]: Why have we added this?

File: .circleci/config.yml [hunk 2]
```diff
@@ -64,6 +64,8 @@ jobs:
           no_output_timeout: "30m"
       - store_artifacts:
           path: /go/src/github.com/Azure/aks-engine/_logs
+      - store_artifacts:
```
Comment (tariq1890) [inline .circleci/config.yml]: Why do we need this?  We had removed this as a part of this PR - https://github.com/Azure/acs-engine/pull/4214

File: pkg/api/types.go [hunk 3]
```diff
@@ -56,6 +56,7 @@ type ContainerService struct {
 // Properties represents the AKS cluster definition
 type Properties struct {
 	ClusterID               string
+	Name                    string                   `json:"clusterName,omitempty"`
```
Comment (tariq1890) [inline pkg/api/types.go]: This should be `ClusterName`

File: pkg/api/types.go [hunk 4]
```diff
@@ -753,6 +754,16 @@ func (p *Properties) getAgentPoolIndexByName(name string) int {
 	return index
 }
 
+// GetPrefix returns resource prefix based on Properties
+func (p *Properties) GetPrefix() (ret string) {
```
Comment (tariq1890) [inline pkg/api/types.go]: GetPrefix is too vague, let's have something specific. `GetClusterPrefix` would work IMO. Also open to other names :)

**Comment by tariq1890** [comment]
@sylr Done!

File: pkg/helpers/helpers.go (hunk 1 above)
Comment (sylr) [reply pkg/helpers/helpers.go]: That's a mistake

File: .circleci/config.yml (hunk 2 above)
Comment (sylr) [reply .circleci/config.yml]: I added this for debug purposes.

Why did you remove it ?

File: pkg/api/types.go (hunk 3 above)
Comment (sylr) [reply pkg/api/types.go]: All right

File: pkg/api/types.go (hunk 4 above)
Comment (sylr) [reply pkg/api/types.go]: `GetClusterName` ?

File: test/e2e/azure/cli.go [hunk 5]
```diff
@@ -212,7 +211,8 @@ func (a *Account) CreateDeployment(name string, e *engine.Engine) error {
 		"--name", d.Name,
 		"--resource-group", a.ResourceGroup.Name,
 		"--template-file", e.Config.GeneratedTemplatePath,
-		"--parameters", e.Config.GeneratedParametersPath)
+		"--parameters", e.Config.GeneratedParametersPath,
```
Comment (tariq1890) [inline test/e2e/azure/cli.go]: Assuming these were added for debugging and not meant to be included in the PR.

**Comment by jackfrancis** [comment]
@sylr Can you kindly add a detailed description of the changes in the PR?

**Comment by jackfrancis** [comment]
Thanks for this @sylr 

The change as-is is breaking. E.g.:

Before:

`"masterVMNamePrefix": "k8s-master-63250626-",`

After:

`"masterVMNamePrefix": "k8s-63250626-master-",`

We should make as a requirement of this change that it be non-breaking when the `clusterName` property is not used.

**Comment by sylr** [comment]
@jackfrancis I disagree on this.

We shouldn't have resources with different name patterns.

File: .circleci/config.yml (hunk 2 above)
Comment (CecileRobertMichon) [reply .circleci/config.yml]: let's revert adding this back - circle ci is not used anymore

Comment (sylr) [reply .circleci/config.yml]: ok

**Comment by jackfrancis** [comment]
@sylr I'm not sure what you mean. The important thing, to preserve backwards-compatibility, is that any new change like this not effect previous usage patterns. In this specific case, we need to make sure that the vm name generation implementation produces the same output as it did before when this new feature is not enabled.

**Pull request closed without merging**